lib/cn/route.c

```diff
@@ -0,0 +1,247 @@
+    if (cc < 1)
+        return NULL;
+
+    n = sscanf(buf, "%u%u%ms%u", &fanout, &pfxlen, &fmt, &skip);
+
+    if (n < 3 || fanout != cp->fanout || pfxlen != cp->pfx_len || skip < 1) {
+        log_err("fanout (%u vs %u), pfxlen (%u vs %u), skip %u",
+                fanout, cp->fanout, pfxlen, cp->pfx_len, skip);
+        return NULL;
+    }
+
+    if (n < 4 || skip < 1)
+        skip = 1;
+
+    if (!strcmp(fmt, "binary")) {
+        if (pfxlen > sizeof(binkeybuf)) {
+            log_err("pfxlen %u > binkeybuf %zu", pfxlen, sizeof(binkeybuf));
+            return NULL;
+        }
+
+        free(fmt);
+        fmt = NULL;
+    }
+
+    sz = sizeof(*map) + sizeof(map->rtm_nodev[0]) * fanout;
+
+    map = aligned_alloc(4096, roundup(sz, 4096));
+    if (!map) {
+        free(fmt);
+        return NULL;
+    }
+
+    memset(map, 0, sz);
+    map->rtm_fanout = fanout;
+    map->rtm_pfxlen = pfxlen;
+
+    for (uint i = 0; i < fanout; ++i) {
+        struct route_node *node = map->rtm_nodev + i;
+        struct route_node *dup;
+        uint fmtarg;
+        int n;
+
+        fmtarg = (i + 1) * skip - 1;
+        node->rtn_keylen = pfxlen;
+        node->rtn_child = i;
+
+        if (fmt) {
+            n = snprintf((char *)node->rtn_keybuf, sizeof(node->rtn_keybuf), fmt, fmtarg);
+
+            if (n < 1 || n >= sizeof(node->rtn_keybuf)) {
+                log_err("overflow %u: n %d, pfxlen %u, fmt [%s], fmtarg %u",
+                        i, n, pfxlen, fmt, fmtarg);
+                abort();
+            }
+
+            if (n != pfxlen) {
+                log_err("skipping %u: n %d, pfxlen %u, fmt [%s], fmtarg %u",
+                        i, n, pfxlen, fmt, fmtarg);
+                continue;
+            }
+        } else {
+            binkeybuf = cpu_to_be64(fmtarg);
+
+            memcpy(node->rtn_keybuf, (char *)&binkeybuf + (sizeof(binkeybuf) - pfxlen), pfxlen);
+        }
+
+        dup = route_map_insert(map, node);
+        if (dup) {
+            log_err("dup route %u: child %u, pfxlen %u, fmt [%s], fmtarg %u",
+                    i, dup->rtn_child, pfxlen, fmt ?: "binary", fmtarg);
+
+            node->rtn_next = map->rtm_free;
+            map->rtm_free = node;
+        }
+    }
+
+    free(fmt);
+
+    if (!rb_first(&map->rtm_root)) {
+        free(map);
+        return NULL;
+    }
+
+    if (1) {
+        struct rb_node *node;
+
+        node = rb_first(&map->rtm_root);
+
+        while (node) {
+            struct route_node *this = rb_entry(node, struct route_node, rtn_node);
+            char buf[1024];
+            int n = 0;
+
+            for (uint i = 0; i < this->rtn_keylen; ++i)
+                n += snprintf(buf + n, sizeof(buf) - n, " %02x", this->rtn_keybuf[i]);
+
+            log_err("%3u %2u:%s", this->rtn_child, this->rtn_keylen, buf);
+
+            node = rb_next(node);
+        }
+    }
+
+
+    return map;
+}
+
+void
+route_map_destroy(struct route_map *map)
+{
+    if (!map)
```

Comment:
This check can be removed as free(NULL) should be fine..

Reply:
Yep.  I used to do the "dump the edge keys" here, but then I moved it into create...